### ISSUE TITLE
Run Locally

### ISSUE THREAD

**Issue by sirinath** [microsoft/python-in-excel#11]
Please ad an option to run locally with the installed Python distribution.

**Comment by stonebig** [comment]
Maybe then there is a big warm-up cost ?

**Comment by richardwesthaver** [comment]
I’d like to second this feature request - Excel -> Python is a great pathway for getting folks interested in programming and this could be a great tool for education. It would be awesome to have an ‘escape hatch’ toggle though which doesn’t rely on O365 services and would make for a very smooth transition to programming outside of excel workbooks.

Also there are some orgs which may be interested in this feature for security/legal reasons - where cloud compute resources need to go through some lengthy audits.

Cheers,
Richard

**Comment by sirinath** [comment]
Maybe the Python can be integrated with PowerFX (https://github.com/microsoft/Power-Fx) too.

**Comment by jflam** [comment]
> I’d like to second this feature request - Excel -> Python is a great pathway for getting folks interested in programming and this could be a great tool for education. It would be awesome to have an ‘escape hatch’ toggle though which doesn’t rely on O365 services and would make for a very smooth transition to programming outside of excel workbooks.

We've definitely thought about the possibilities in education here and what we can do with potentially WASM hosted local execution. But that tech isn't mature enough yet for products like Excel, but we are monitoring (and have MSFT employees actively working on Python in WASM) this closely.

Do you have any examples of a scenario where you would start from an Excel workbook and then transition the Python code to something running outside of Excel? Would love to learn more about this.

**Comment by GaryT25** [comment]
> I’d like to second this feature request - Excel -> Python is a great pathway for getting folks interested in programming and this could be a great tool for education. It would be awesome to have an ‘escape hatch’ toggle though which doesn’t rely on O365 services and would make for a very smooth transition to programming outside of excel workbooks.
> 
> 
> 
> Also there are some orgs which may be interested in this feature for security/legal reasons - where cloud compute resources need to go through some lengthy audits.
> 
> 
> 
> Cheers,
> 
> Richard
> 
> 

I agree. A lot of organizations do not like to add cloud resources when not absolutely necessary, and this feature would be amazing using just your local machine with python installed!

**Comment by stonebig** [comment]
On a test sample of 1, the main complaint of Python teachers in a Classroom is the internet Bandwith & reliability. You can't rely on internet for your lessons.

**Comment by elubarsky-hs** [comment]
Running locally would likely have reduced CO2 impact compared with a cloud solution as data doesn't have to be transferred to/from the cloud (especially if data volume is large)? And it would be much better when internet access is slow or expensive..

**Comment by jflam** [comment]
I tried to answer this question in this tweet: https://twitter.com/john_lam/status/1694480333184569400. Also copying and pasting it into this thread as well. Hope this helps clarify the design constraints that we had for this feature!

A common question that has come up is why don't we run Python locally? There are three main reasons:

1. Running Python securely on a local machine is a really hard problem. We treat all Python code in the workbook as untrusted, so we execute it in a hypervisor-isolated container on Azure that does not have any outbound network access. Python code and the data that it operates on is sent to be executed in the container. The Microsoft-licensed Python environment in the container is provided by Anaconda and was prepared using their stringent security practices as documented here: https://docs.anaconda.com/free/anaconda/reference/security/

2. Sharing Excel workbooks with others is a really important scenario. We wanted to ensure that Python code in a workbook that you share will behave as you intended when someone else opens it; they wouldn't first need to install Python first.

3. We need to ensure that the Python in Excel feature always works for our customers. The value of Python is in its ecosystem of libraries, not just in providing a Python interpreter. But managing a local Python environment is challenging even for the most experienced developers. By running on Azure, we remove the need for users or their systems administrators to have to maintain a local installation of Python on every machine that uses the feature in their organization.

**Comment by DrEntropy** [comment]
> I tried to answer this question in this tweet: https://twitter.com/john_lam/status/1694480333184569400. Also copying and pasting it into this thread as well. Hope this helps clarify the design constraints that we had for this feature!
 
Ah, I had not thought of it in that way.  This was very helpful!  Still would love to see an offline mode ('unsafe' or using local docker containers) in the future, but now I think I understand why you went down this road first.

**Comment by sirinath** [comment]
You can have the option do run the container locally.

**Comment by jflam** [comment]
Local container runtimes do not provide sufficient isolation today. It's also true that firewalls / NAT rules could block access as well. What we shipped is a tradeoff that we had to make to ship something sooner rather than later once all issues could be resolved. It's a hard problem!

**Comment by ismasanchez** [comment]
While I understand the security concerns, the user/organization should be the ones that weights the risk and allow the possibility to run the code locally. My proposal is to provide some option to configure the environment where python code should run. This way, personal or professionals could use all its potential, and large organizations or public institutions (education, etc...) could limit its usage to mitigate risks.

**Comment by Julrob199** [comment]
> A common question that has come up is why don't we run Python locally? There are three main reasons:...

For me that also makes sense. So a cloud-default and fallback switch for local would be nice.

**Another use case, why cloud is currently not an option**: I created a business case with python formulas to reduce the formulas and simplify the case. But it is too slow to calculate. I used up my quota after one or two days. While I would still able to upgrade to more cloud calculation power, my coworkers simulating with the case, are not.

because the case is so slow calculate[^1], I had to turn back to regular excel formulas, which is sad, as python has much more flexibility and reusability power with functions, ...

I was highly enthusiastic about python in excel and made lots of advertisement in the department. All coworkers were sceptical (due to lack of skills in python, scepticism of innovation, ...). Now it turns out they are right, because the excel was not usable for simulation as it was too slow.

**Possible solution**: It would be really cool to either have a local calculation option or make the default cloud calculation so fast, that regular operations in python are the same speed like local excel formulas. A premium package for cloud operation would then only be necessecary for high computational demand (like machine learning, ...). But currently python in excel is out of scope now in our department.

[^1]: every change of number requires waiting for 10-15 seconds to recalculate. My business case even did not have complex python functions (machine learning or anything like this, just normal mathematical pandas operations). I maybe had 20 cells with python formulas, but still it was waay to slow to be calculated and be usable in production readiness. I even tried high performant numpy calculations, but still same waiting length. So there seems to be a big overhead of waiting time, which does not depend on the code itself, but initilizing the python, sending the data, ...

**Comment by salabim** [comment]
Try xlwings lite. It is so much better than the Microsoft product.
